### PR TITLE
Cpp20 bit manipulations

### LINKED ISSUE #543: Make use of C++20 bit manipulation features
We are currently using compiler builtins:
- `__builtin_clz`
- `__builtin_clzll`
- `__builtin_ctz`
- `__builtin_ctzll`
- `__builtin_popcount`
- `__builtin_popcountll`

These can be replaced with the new C++20 features proposed (see https://en.cppreference.com/w/cpp/numeric).

For an idea how to write such a wrapper, check out `lib/fizzy/cxx20/span.hpp`.

### PR DESCRIPTION
**Pull request by herobank110**
Closes #543.

### CODE REVIEW

File: lib/fizzy/cxx20/bit.hpp [hunk 1]
```diff
@@ -24,3 +25,45 @@ template <class To, class From>
...
+}
+}  // namespace fizzy
+
+#else
+
+namespace fizzy
+{
+inline uint32_t countl_zero(uint32_t value) noexcept
+{
+    // NOTE: Wasm specifies this case, but C/C++ intrinsic leaves it as undefined.
```
Comment (axic) [inline lib/fizzy/cxx20/bit.hpp:54]: Here we are not copying Wasm behaviour, but C++20 behaviour. These functions must 100% match how `countl_zero` behaves in C++20.

File: lib/fizzy/cxx20/bit.hpp [hunk 2]
```diff
@@ -24,3 +25,45 @@ template <class To, class From>
     return dst;
 }
 }  // namespace fizzy
+
+#if __cplusplus > 201703L
```
Comment (axic) [inline lib/fizzy/cxx20/bit.hpp:29]: Please use the appropriate `__cpp_lib_*` macro here.

File: lib/fizzy/cxx20/bit.hpp (hunk 1 above)
Comment (herobank110) [reply lib/fizzy/cxx20/bit.hpp]: In the original execute.cpp versions, the return type was uint32_t and uint64_t for these bit counting functions. https://github.com/wasmx/fizzy/blob/master/lib/fizzy/execute.cpp#L260

But c++ 20 standard library and c++ intrinsics both have them returning int. Which version should I do?

**Comment by axic** [comment]
Please take #551 as a template how to do this. And make sure to rebase this branch on master.

**Comment by axic** [comment]
@herobank110 are you still pursuing this?

**Comment by herobank110** [comment]
sorry for the lack of work recently. I am still interested in this issue

**Comment by axic** [comment]
There seems to be some issue with the commit history. Can you rebase this top on master? (e.g. `git rebase master`)

**Comment by herobank110** [comment]
I did the rebase but not sure if it is correct.

**Comment by axic** [comment]
It doesn't seem to be correct, there are still over 140 commits in this PR.

You could do the following:
```sh
git remote add upstream git@github.com:wasmx/fizzy.git
git fetch upstream
git rebase upstream/master
```

File: lib/fizzy/cxx20/bit.hpp [hunk 3]
```diff
@@ -62,6 +83,35 @@ constexpr int popcount(uint64_t x) noexcept
...
+constexpr int countl_zero(uint32_t value) noexcept
+{
+    if (value == 0)
+        return 32;
+    return __builtin_clz(value);
+}
+
+constexpr int countl_zero(uint64_t value) noexcept
+{
+    // NOTE: C++20 specifies this case, but C/C++ intrinsic leaves it as undefined.
```
Comment (axic) [inline lib/fizzy/cxx20/bit.hpp:96]:
```suggestion
```

**Comment by axic** [comment]
Thank you, this was merged last year (under #689).

**Pull request closed without merging**